File: golang/internal/storage/aws.go
// Package storage provides the AWS S3 gateway backend for BleepStore.
//
// The AWS gateway backend proxies all data operations to an upstream AWS S3
// bucket via the AWS SDK for Go v2. Metadata stays in local SQLite -- this
// backend handles raw bytes only.
//
// Key mapping:
//
//	Objects:  {prefix}{bleepstore_bucket}/{key}
//	Parts:    {prefix}.parts/{upload_id}/{part_number}
//
// Credentials are resolved via the standard AWS credential chain
// (env vars, ~/.aws/credentials, IAM role, etc.).
package storage

import (
	"bytes"
	"context"
	"crypto/md5"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"strings"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/aws/smithy-go"
)

// S3API defines the subset of the AWS S3 client interface that the gateway
// backend uses. This allows mocking in tests.
type S3API interface {
	PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error)
	GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error)
	DeleteObject(ctx context.Context, params *s3.DeleteObjectInput, optFns ...func(*s3.Options)) (*s3.DeleteObjectOutput, error)
	DeleteObjects(ctx context.Context, params *s3.DeleteObjectsInput, optFns ...func(*s3.Options)) (*s3.DeleteObjectsOutput, error)
	CopyObject(ctx context.Context, params *s3.CopyObjectInput, optFns ...func(*s3.Options)) (*s3.CopyObjectOutput, error)
	HeadObject(ctx context.Context, params *s3.HeadObjectInput, optFns ...func(*s3.Options)) (*s3.HeadObjectOutput, error)
	HeadBucket(ctx context.Context, params *s3.HeadBucketInput, optFns ...func(*s3.Options)) (*s3.HeadBucketOutput, error)
	CreateMultipartUpload(ctx context.Context, params *s3.CreateMultipartUploadInput, optFns ...func(*s3.Options)) (*s3.CreateMultipartUploadOutput, error)
	UploadPart(ctx context.Context, params *s3.UploadPartInput, optFns ...func(*s3.Options)) (*s3.UploadPartOutput, error)
	UploadPartCopy(ctx context.Context, params *s3.UploadPartCopyInput, optFns ...func(*s3.Options)) (*s3.UploadPartCopyOutput, error)
	CompleteMultipartUpload(ctx context.Context, params *s3.CompleteMultipartUploadInput, optFns ...func(*s3.Options)) (*s3.CompleteMultipartUploadOutput, error)
	AbortMultipartUpload(ctx context.Context, params *s3.AbortMultipartUploadInput, optFns ...func(*s3.Options)) (*s3.AbortMultipartUploadOutput, error)
	ListObjectsV2(ctx context.Context, params *s3.ListObjectsV2Input, optFns ...func(*s3.Options)) (*s3.ListObjectsV2Output, error)
}

// AWSGatewayBackend implements the StorageBackend interface by proxying
// storage operations to an upstream Amazon S3 bucket. This allows BleepStore
// to act as an S3-compatible gateway in front of native AWS S3.
//
// All BleepStore buckets/objects are stored under a single upstream S3 bucket
// with a key prefix to namespace them.
type AWSGatewayBackend struct {
	// Bucket is the upstream S3 bucket name.
	Bucket string
	// Region is the AWS region of the upstream bucket.
	Region string
	// Prefix is the key prefix for all objects in the upstream bucket.
	Prefix string
	// client is the AWS S3 client (satisfying S3API interface).
	client S3API
}

// NewAWSGatewayBackend creates a new AWSGatewayBackend configured to proxy
// to the specified S3 bucket in the given region. It initializes the AWS SDK
// client using the default credential chain, with optional overrides for
// custom endpoint, path-style addressing, and static credentials.
func NewAWSGatewayBackend(ctx context.Context, bucket, region, prefix, endpointURL string, usePathStyle bool, accessKeyID, secretAccessKey string) (*AWSGatewayBackend, error) {
	var loadOpts []func(*awsconfig.LoadOptions) error
	loadOpts = append(loadOpts, awsconfig.WithRegion(region))

	// Use static credentials if provided, otherwise fall back to default chain.
	if accessKeyID != "" && secretAccessKey != "" {
		loadOpts = append(loadOpts, awsconfig.WithCredentialsProvider(
			credentials.NewStaticCredentialsProvider(accessKeyID, secretAccessKey, ""),
		))
	}

	cfg, err := awsconfig.LoadDefaultConfig(ctx, loadOpts...)
	if err != nil {
		return nil, fmt.Errorf("loading AWS config: %w", err)
	}

	// Build S3 client options for custom endpoint and path-style.
	var s3Opts []func(*s3.Options)
	if endpointURL != "" {
		s3Opts = append(s3Opts, func(o *s3.Options) {
			o.BaseEndpoint = aws.String(endpointURL)
		})
	}
	if usePathStyle {
		s3Opts = append(s3Opts, func(o *s3.Options) {
			o.UsePathStyle = true
		})
	}

	client := s3.NewFromConfig(cfg, s3Opts...)

	b := &AWSGatewayBackend{
		Bucket: bucket,
		Region: region,
		Prefix: prefix,
		client: client,
	}

	// Verify the upstream bucket is accessible.
	_, err = client.HeadBucket(ctx, &s3.HeadBucketInput{
		Bucket: aws.String(bucket),
	})
	if err != nil {
		return nil, fmt.Errorf("cannot access upstream S3 bucket %q: %w", bucket, err)
	}

	slog.Info("AWS gateway backend initialized", "bucket", bucket, "region", region, "prefix", prefix)
	return b, nil
}

// NewAWSGatewayBackendWithClient creates an AWSGatewayBackend with a
// pre-configured S3 client. This is primarily used for testing with mock
// clients.
func NewAWSGatewayBackendWithClient(bucket, region, prefix string, client S3API) *AWSGatewayBackend {
	return &AWSGatewayBackend{
		Bucket: bucket,
		Region: region,
		Prefix: prefix,
		client: client,
	}
}

// s3Key maps a BleepStore bucket/key to an upstream S3 key.
func (b *AWSGatewayBackend) s3Key(bucket, key string) string {
	return b.Prefix + bucket + "/" + key
}

// partKey maps a multipart part to an upstream S3 key.
func (b *AWSGatewayBackend) partKey(uploadID string, partNumber int) string {
	return fmt.Sprintf("%s.parts/%s/%d", b.Prefix, uploadID, partNumber)
}

// PutObject uploads object data to the upstream S3 bucket. It reads all data,
// computes MD5 locally for a consistent ETag, then uploads to S3.
func (b *AWSGatewayBackend) PutObject(ctx context.Context, bucket, key string, reader io.Reader, size int64) (int64, string, error) {
	s3key := b.s3Key(bucket, key)

	// Read all data to compute MD5 locally. AWS may return different ETags
	// when server-side encryption is enabled, so we compute our own.
	data, err := io.ReadAll(reader)
	if err != nil {
		return 0, "", fmt.Errorf("reading object data: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))

	_, err = b.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:        aws.String(b.Bucket),
		Key:           aws.String(s3key),
		Body:          bytes.NewReader(data),
		ContentLength: aws.Int64(int64(len(data))),
	})
	if err != nil {
		return 0, "", fmt.Errorf("uploading to S3: %w", err)
	}

	return int64(len(data)), etag, nil
}

// GetObject retrieves object data from the upstream S3 bucket.
// Returns the data stream, the object size, and an empty ETag (metadata store
// holds the authoritative ETag). The caller is responsible for closing the
// returned ReadCloser.
func (b *AWSGatewayBackend) GetObject(ctx context.Context, bucket, key string) (io.ReadCloser, int64, string, error) {
	s3key := b.s3Key(bucket, key)

	resp, err := b.client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(b.Bucket),
		Key:    aws.String(s3key),
	})
	if err != nil {
		if isAWSNotFound(err) {
			return nil, 0, "", fmt.Errorf("object not found: %s/%s", bucket, key)
		}
		return nil, 0, "", fmt.Errorf("getting object from S3: %w", err)
	}

	var objectSize int64
	if resp.ContentLength != nil {
		objectSize = *resp.ContentLength
	}

	return resp.Body, objectSize, "", nil
}

// DeleteObject removes an object from the upstream S3 bucket.
// Idempotent: S3 DeleteObject does not error on missing keys.
func (b *AWSGatewayBackend) DeleteObject(ctx context.Context, bucket, key string) error {
	s3key := b.s3Key(bucket, key)

	_, err := b.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(b.Bucket),
		Key:    aws.String(s3key),
	})
	if err != nil {
		return fmt.Errorf("deleting object from S3: %w", err)
	}
	return nil
}

// CopyObject copies an object within the upstream S3 bucket using AWS
// server-side copy. Returns the new ETag computed locally by downloading
// the source to ensure consistent ETags.
func (b *AWSGatewayBackend) CopyObject(ctx context.Context, srcBucket, srcKey, dstBucket, dstKey string) (string, error) {
	srcS3Key := b.s3Key(srcBucket, srcKey)
	dstS3Key := b.s3Key(dstBucket, dstKey)

	copySource := b.Bucket + "/" + srcS3Key

	resp, err := b.client.CopyObject(ctx, &s3.CopyObjectInput{
		Bucket:     aws.String(b.Bucket),
		Key:        aws.String(dstS3Key),
		CopySource: aws.String(copySource),
	})
	if err != nil {
		if isAWSNotFound(err) {
			return "", fmt.Errorf("source object not found: %s/%s", srcBucket, srcKey)
		}
		return "", fmt.Errorf("copying object in S3: %w", err)
	}

	// Extract ETag from CopyObjectResult, stripping quotes.
	etag := ""
	if resp.CopyObjectResult != nil && resp.CopyObjectResult.ETag != nil {
		etag = strings.Trim(*resp.CopyObjectResult.ETag, `"`)
	}

	return fmt.Sprintf(`"%s"`, etag), nil
}

// PutPart stores a multipart upload part as a temporary S3 object.
// Parts are stored at {prefix}.parts/{upload_id}/{part_number}.
// Computes MD5 locally for a consistent ETag.
func (b *AWSGatewayBackend) PutPart(ctx context.Context, bucket, key, uploadID string, partNumber int, reader io.Reader, size int64) (string, error) {
	pk := b.partKey(uploadID, partNumber)

	// Read all data to compute MD5 locally.
	data, err := io.ReadAll(reader)
	if err != nil {
		return "", fmt.Errorf("reading part data: %w", err)
	}

	h := md5.New()
	h.Write(data)
	etag := fmt.Sprintf(`"%x"`, h.Sum(nil))

	_, err = b.client.PutObject(ctx, &s3.PutObjectInput{
		Bucket:        aws.String(b.Bucket),
		Key:           aws.String(pk),
		Body:          bytes.NewReader(data),
		ContentLength: aws.Int64(int64(len(data))),
	})
	if err != nil {
		return "", fmt.Errorf("uploading part to S3: %w", err)
	}

	return etag, nil
}

// AssembleParts assembles uploaded parts into the final object using AWS
// native multipart upload with UploadPartCopy for server-side copy.
//
// For a single part, uses CopyObject directly. For multiple parts, creates
// a native AWS multipart upload and uses UploadPartCopy for server-side
// assembly (no data download). Falls back to download + re-upload if
// UploadPartCopy fails with EntityTooSmall.
//
// Returns the composite ETag.
func (b *AWSGatewayBackend) AssembleParts(ctx context.Context, bucket, key, uploadID string, partNumbers []int) (string, error) {
	finalKey := b.s3Key(bucket, key)

	if len(partNumbers) == 1 {
		// Single part: direct copy.
		pk := b.partKey(uploadID, partNumbers[0])
		copySource := b.Bucket + "/" + pk

		resp, err := b.client.CopyObject(ctx, &s3.CopyObjectInput{
			Bucket:     aws.String(b.Bucket),
			Key:        aws.String(finalKey),
			CopySource: aws.String(copySource),
		})
		if err != nil {
			return "", fmt.Errorf("copying single part to final object: %w", err)
		}

		etag := ""
		if resp.CopyObjectResult != nil && resp.CopyObjectResult.ETag != nil {
			etag = strings.Trim(*resp.CopyObjectResult.ETag, `"`)
		}
		return fmt.Sprintf(`"%s"`, etag), nil
	}

	// Multiple parts: native AWS multipart upload with server-side copy.
	createResp, err := b.client.CreateMultipartUpload(ctx, &s3.CreateMultipartUploadInput{
		Bucket: aws.String(b.Bucket),
		Key:    aws.String(finalKey),
	})
	if err != nil {
		return "", fmt.Errorf("creating AWS multipart upload: %w", err)
	}
	awsUploadID := aws.ToString(createResp.UploadId)

	var completedParts []types.CompletedPart
	abortOnError := func() {
		_, abortErr := b.client.AbortMultipartUpload(ctx, &s3.AbortMultipartUploadInput{
			Bucket:   aws.String(b.Bucket),
			Key:      aws.String(finalKey),
			UploadId: aws.String(awsUploadID),
		})
		if abortErr != nil {
			slog.Warn("Failed to abort AWS multipart upload", "upload_id", awsUploadID, "error", abortErr)
		}
	}

	for idx, pn := range partNumbers {
		awsPartNumber := int32(idx + 1) // AWS part numbers are 1-indexed
		pk := b.partKey(uploadID, pn)
		copySource := b.Bucket + "/" + pk

		copyResp, copyErr := b.client.UploadPartCopy(ctx, &s3.UploadPartCopyInput{
			Bucket:     aws.String(b.Bucket),
			Key:        aws.String(finalKey),
			UploadId:   aws.String(awsUploadID),
			PartNumber: aws.Int32(awsPartNumber),
			CopySource: aws.String(copySource),
		})

		var partETag string
		if copyErr != nil {
			// Check if it's EntityTooSmall -- fall back to download + re-upload.
			if isAWSEntityTooSmall(copyErr) {
				getResp, getErr := b.client.GetObject(ctx, &s3.GetObjectInput{
					Bucket: aws.String(b.Bucket),
					Key:    aws.String(pk),
				})
				if getErr != nil {
					abortOnError()
					return "", fmt.Errorf("downloading part %d for fallback upload: %w", pn, getErr)
				}
				partData, readErr := io.ReadAll(getResp.Body)
				getResp.Body.Close()
				if readErr != nil {
					abortOnError()
					return "", fmt.Errorf("reading part %d data: %w", pn, readErr)
				}

				uploadResp, uploadErr := b.client.UploadPart(ctx, &s3.UploadPartInput{
					Bucket:     aws.String(b.Bucket),
					Key:        aws.String(finalKey),
					UploadId:   aws.String(awsUploadID),
					PartNumber: aws.Int32(awsPartNumber),
					Body:       bytes.NewReader(partData),
				})
				if uploadErr != nil {
					abortOnError()
					return "", fmt.Errorf("uploading part %d fallback: %w", pn, uploadErr)
				}
				partETag = aws.ToString(uploadResp.ETag)
			} else {
				abortOnError()
				return "", fmt.Errorf("copying part %d: %w", pn, copyErr)
			}
		} else {
			if copyResp.CopyPartResult != nil && copyResp.CopyPartResult.ETag != nil {
				partETag = *copyResp.CopyPartResult.ETag
			}
		}

		completedParts = append(completedParts, types.CompletedPart{
			ETag:       aws.String(partETag),
			PartNumber: aws.Int32(awsPartNumber),
		})
	}

	completeResp, err := b.client.CompleteMultipartUpload(ctx, &s3.CompleteMultipartUploadInput{
		Bucket:   aws.String(b.Bucket),
		Key:      aws.String(finalKey),
		UploadId: aws.String(awsUploadID),
		MultipartUpload: &types.CompletedMultipartUpload{
			Parts: completedParts,
		},
	})
	if err != nil {
		abortOnError()
		return "", fmt.Errorf("completing AWS multipart upload: %w", err)
	}

	etag := ""
	if completeResp.ETag != nil {
		etag = strings.Trim(*completeResp.ETag, `"`)
	}
	return fmt.Sprintf(`"%s"`, etag), nil
}

// DeleteParts removes all temporary part objects for a multipart upload.
// Lists objects under .parts/{upload_id}/ and batch-deletes them.
func (b *AWSGatewayBackend) DeleteParts(ctx context.Context, bucket, key, uploadID string) error {
	prefix := b.Prefix + ".parts/" + uploadID + "/"

	// List all part objects under this upload ID.
	for {
		listResp, err := b.client.ListObjectsV2(ctx, &s3.ListObjectsV2Input{
			Bucket: aws.String(b.Bucket),
			Prefix: aws.String(prefix),
		})
		if err != nil {
			return fmt.Errorf("listing parts for upload %s: %w", uploadID, err)
		}

		if len(listResp.Contents) == 0 {
			break
		}

		// Build the delete objects list.
		var objects []types.ObjectIdentifier
		for _, obj := range listResp.Contents {
			objects = append(objects, types.ObjectIdentifier{
				Key: obj.Key,
			})
		}

		_, err = b.client.DeleteObjects(ctx, &s3.DeleteObjectsInput{
			Bucket: aws.String(b.Bucket),
			Delete: &types.Delete{
				Objects: objects,
				Quiet:   aws.Bool(true),
			},
		})
		if err != nil {
			return fmt.Errorf("batch-deleting parts for upload %s: %w", uploadID, err)
		}

		if !aws.ToBool(listResp.IsTruncated) {
			break
		}
	}

	return nil
}

// CreateBucket is a no-op for the AWS gateway backend. All BleepStore buckets
// share a single upstream S3 bucket with key prefixes, so there is nothing
// to create on the AWS side.
func (b *AWSGatewayBackend) CreateBucket(ctx context.Context, bucket string) error {
	// No-op: BleepStore buckets are mapped to key prefixes within the upstream
	// bucket, not to actual AWS S3 buckets.
	return nil
}

// DeleteBucket is a no-op for the AWS gateway backend. Bucket data is scoped
// by key prefix, so there is nothing to delete on the AWS side. The metadata
// store handles the actual bucket record deletion.
func (b *AWSGatewayBackend) DeleteBucket(ctx context.Context, bucket string) error {
	// No-op: same as CreateBucket. Objects under the prefix will be deleted
	// individually by DeleteObject calls.
	return nil
}

// ObjectExists checks whether an object exists in the upstream S3 bucket.
func (b *AWSGatewayBackend) ObjectExists(ctx context.Context, bucket, key string) (bool, error) {
	s3key := b.s3Key(bucket, key)

	_, err := b.client.HeadObject(ctx, &s3.HeadObjectInput{
		Bucket: aws.String(b.Bucket),
		Key:    aws.String(s3key),
	})
	if err != nil {
		if isAWSNotFound(err) {
			return false, nil
		}
		return false, fmt.Errorf("checking object existence in S3: %w", err)
	}
	return true, nil
}

// HealthCheck verifies that the upstream S3 bucket is accessible.
func (b *AWSGatewayBackend) HealthCheck(ctx context.Context) error {
	_, err := b.client.HeadBucket(ctx, &s3.HeadBucketInput{
		Bucket: aws.String(b.Bucket),
	})
	return err
}

// isAWSNotFound checks if an AWS error is a 404/NoSuchKey/NotFound error.
func isAWSNotFound(err error) bool {
	var apiErr smithy.APIError
	if errors.As(err, &apiErr) {
		code := apiErr.ErrorCode()
		if code == "NoSuchKey" || code == "NotFound" || code == "404" || code == "NoSuchBucket" {
			return true
		}
	}
	// Also check for types.NoSuchKey.
	var noSuchKey *types.NoSuchKey
	if errors.As(err, &noSuchKey) {
		return true
	}
	// Check HTTP status code via ResponseError.
	var respErr interface{ HTTPStatusCode() int }
	if errors.As(err, &respErr) {
		if respErr.HTTPStatusCode() == 404 {
			return true
		}
	}
	return false
}

// isAWSEntityTooSmall checks if an AWS error is an EntityTooSmall error.
func isAWSEntityTooSmall(err error) bool {
	var apiErr smithy.APIError
	if errors.As(err, &apiErr) {
		return apiErr.ErrorCode() == "EntityTooSmall"
	}
	return false
}

// Ensure AWSGatewayBackend implements StorageBackend at compile time.
var _ StorageBackend = (*AWSGatewayBackend)(nil)

File: python/src/bleepstore/storage/aws.py
"""AWS S3 gateway storage backend for BleepStore.

Proxies all data operations to an upstream AWS S3 bucket via aiobotocore.
Metadata stays in local SQLite — this backend handles raw bytes only.

Key mapping:
    Objects:  {prefix}{bleepstore_bucket}/{key}
    Parts:    {prefix}.parts/{upload_id}/{part_number}

Credentials are resolved via the standard AWS credential chain
(env vars, ~/.aws/credentials, IAM role, etc.).
"""

import hashlib
import logging
from collections.abc import AsyncIterator

from aiobotocore.session import AioSession
from botocore.exceptions import ClientError

logger = logging.getLogger(__name__)

# Streaming chunk size: 64 KB (matches local backend)
_CHUNK_SIZE = 64 * 1024


class AWSGatewayBackend:
    """Storage backend that proxies to a real AWS S3 bucket.

    All BleepStore buckets/objects are stored under a single upstream
    S3 bucket with a key prefix to namespace them.

    Attributes:
        bucket_name: The upstream AWS S3 bucket name.
        region: The AWS region for the bucket.
        prefix: Key prefix for all objects in the upstream bucket.
    """

    def __init__(
        self,
        bucket_name: str,
        region: str = "us-east-1",
        prefix: str = "",
        endpoint_url: str = "",
        use_path_style: bool = False,
        access_key_id: str = "",
        secret_access_key: str = "",
    ) -> None:
        self.bucket_name = bucket_name
        self.region = region
        self.prefix = prefix
        self.endpoint_url = endpoint_url
        self.use_path_style = use_path_style
        self.access_key_id = access_key_id
        self.secret_access_key = secret_access_key
        self._session = AioSession()
        self._client = None
        self._client_ctx = None

    def _s3_key(self, bucket: str, key: str) -> str:
        """Map a BleepStore bucket/key to an upstream S3 key."""
        return f"{self.prefix}{bucket}/{key}"

    def _part_key(self, upload_id: str, part_number: int) -> str:
        """Map a multipart part to an upstream S3 key."""
        return f"{self.prefix}.parts/{upload_id}/{part_number}"

    async def init(self) -> None:
        """Create the aiobotocore S3 client and verify the upstream bucket exists.

        Raises:
            ValueError: If the upstream bucket does not exist or is inaccessible.
        """
        # Build client kwargs from config
        client_kwargs: dict = {"region_name": self.region}
        if self.endpoint_url:
            client_kwargs["endpoint_url"] = self.endpoint_url
        if self.use_path_style:
            from botocore.config import Config as BotoConfig
            client_kwargs["config"] = BotoConfig(s3={"addressing_style": "path"})

        # Use explicit credentials if provided, otherwise fall back to chain
        if self.access_key_id and self.secret_access_key:
            session = AioSession()
            session.set_credentials(self.access_key_id, self.secret_access_key)
            self._session = session
        self._client_ctx = self._session.create_client("s3", **client_kwargs)
        self._client = await self._client_ctx.__aenter__()

        # Verify bucket exists
        try:
            await self._client.head_bucket(Bucket=self.bucket_name)
        except ClientError as e:
            code = e.response.get("Error", {}).get("Code", "")
            await self._client_ctx.__aexit__(None, None, None)
            self._client = None
            self._client_ctx = None
            raise ValueError(
                f"Cannot access upstream S3 bucket '{self.bucket_name}': {code}"
            ) from e

        logger.info(
            "AWS gateway backend initialized: bucket=%s region=%s prefix='%s'",
            self.bucket_name,
            self.region,
            self.prefix,
        )

    async def close(self) -> None:
        """Close the aiobotocore client session."""
        if self._client_ctx is not None:
            await self._client_ctx.__aexit__(None, None, None)
            self._client = None
            self._client_ctx = None

    async def put(self, bucket: str, key: str, data: bytes) -> str:
        """Upload an object to the upstream S3 bucket.

        Computes MD5 locally for a consistent ETag (AWS may differ with SSE).

        Returns:
            The hex-encoded MD5 of the stored data.
        """
        s3_key = self._s3_key(bucket, key)
        md5 = hashlib.md5(data).hexdigest()

        await self._client.put_object(
            Bucket=self.bucket_name,
            Key=s3_key,
            Body=data,
        )
        return md5

    async def put_stream(
        self,
        bucket: str,
        key: str,
        stream: AsyncIterator[bytes],
        content_length: int | None = None,
    ) -> tuple[str, int]:
        """Stream-write: collect stream and delegate to put()."""
        chunks = []
        async for chunk in stream:
            chunks.append(chunk)
        data = b"".join(chunks)
        md5_hex = await self.put(bucket, key, data)
        return md5_hex, len(data)

    async def put_part_stream(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        part_number: int,
        stream: AsyncIterator[bytes],
        content_length: int | None = None,
    ) -> tuple[str, int]:
        """Stream-write part: collect stream and delegate to put_part()."""
        chunks = []
        async for chunk in stream:
            chunks.append(chunk)
        data = b"".join(chunks)
        md5_hex = await self.put_part(bucket, key, upload_id, part_number, data)
        return md5_hex, len(data)

    async def get(self, bucket: str, key: str) -> bytes:
        """Download an object from the upstream S3 bucket.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        s3_key = self._s3_key(bucket, key)
        try:
            resp = await self._client.get_object(Bucket=self.bucket_name, Key=s3_key)
        except ClientError as e:
            code = e.response.get("Error", {}).get("Code", "")
            if code in ("NoSuchKey", "404"):
                raise FileNotFoundError(f"Object not found: {bucket}/{key}") from e
            raise

        async with resp["Body"] as stream:
            return await stream.read()

    async def get_stream(
        self, bucket: str, key: str, offset: int = 0, length: int | None = None
    ) -> AsyncIterator[bytes]:
        """Stream an object from the upstream S3 bucket in 64KB chunks.

        Raises:
            FileNotFoundError: If the object does not exist.
        """
        s3_key = self._s3_key(bucket, key)

        kwargs: dict = {"Bucket": self.bucket_name, "Key": s3_key}
        if offset > 0 or length is not None:
            if length is not None:
                end = offset + length - 1
                kwargs["Range"] = f"bytes={offset}-{end}"
            else:
                kwargs["Range"] = f"bytes={offset}-"

        try:
            resp = await self._client.get_object(**kwargs)
        except ClientError as e:
            code = e.response.get("Error", {}).get("Code", "")
            if code in ("NoSuchKey", "404"):
                raise FileNotFoundError(f"Object not found: {bucket}/{key}") from e
            raise

        async with resp["Body"] as stream:
            while True:
                chunk = await stream.read(_CHUNK_SIZE)
                if not chunk:
                    break
                yield chunk

    async def delete(self, bucket: str, key: str) -> None:
        """Delete an object from the upstream S3 bucket.

        Idempotent — S3 delete_object does not error on missing keys.
        """
        s3_key = self._s3_key(bucket, key)
        await self._client.delete_object(Bucket=self.bucket_name, Key=s3_key)

    async def exists(self, bucket: str, key: str) -> bool:
        """Check if an object exists in the upstream S3 bucket."""
        s3_key = self._s3_key(bucket, key)
        try:
            await self._client.head_object(Bucket=self.bucket_name, Key=s3_key)
            return True
        except ClientError as e:
            code = e.response.get("Error", {}).get("Code", "")
            if code in ("404", "NoSuchKey"):
                return False
            raise

    async def put_part(
        self, bucket: str, key: str, upload_id: str, part_number: int, data: bytes
    ) -> str:
        """Store a multipart upload part as a temporary S3 object.

        Parts are stored at {prefix}.parts/{upload_id}/{part_number}.
        Computes MD5 locally for a consistent ETag.

        Returns:
            The hex-encoded MD5 of the part data.
        """
        part_key = self._part_key(upload_id, part_number)
        md5 = hashlib.md5(data).hexdigest()

        await self._client.put_object(
            Bucket=self.bucket_name,
            Key=part_key,
            Body=data,
        )
        return md5

    async def assemble_parts(
        self,
        bucket: str,
        key: str,
        upload_id: str,
        part_numbers: list[int],
    ) -> str:
        """Assemble uploaded parts into the final object using server-side copy.

        For a single part, uses copy_object directly. For multiple parts,
        creates a native AWS multipart upload and uses upload_part_copy
        for server-side assembly (no data download). Falls back to
        download + re-upload if upload_part_copy fails with EntityTooSmall.

        Returns:
            The hex-encoded MD5 of the assembled object (best-effort).
        """
        final_key = self._s3_key(bucket, key)

        if len(part_numbers) == 1:
            # Single part: direct copy
            part_key = self._part_key(upload_id, part_numbers[0])
            resp = await self._client.copy_object(
                Bucket=self.bucket_name,
                Key=final_key,
                CopySource={"Bucket": self.bucket_name, "Key": part_key},
            )
            etag = resp.get("CopyObjectResult", {}).get("ETag", "")
            return etag.strip('"')

        # Multiple parts: native AWS multipart upload with server-side copy
        aws_upload = await self._client.create_multipart_upload(
            Bucket=self.bucket_name, Key=final_key
        )
        aws_upload_id = aws_upload["UploadId"]

        try:
            parts_manifest = []
            for idx, pn in enumerate(part_numbers, 1):
                part_key = self._part_key(upload_id, pn)
                try:
                    copy_resp = await self._client.upload_part_copy(
                        Bucket=self.bucket_name,
                        Key=final_key,
                        UploadId=aws_upload_id,
                        PartNumber=idx,
                        CopySource={
                            "Bucket": self.bucket_name,
                            "Key": part_key,
                        },
                    )
                    etag = copy_resp["CopyPartResult"]["ETag"]
                except ClientError as e:
                    error_code = e.response.get("Error", {}).get("Code", "")
                    if error_code == "EntityTooSmall":
                        # Fallback: download part data and re-upload
                        get_resp = await self._client.get_object(
                            Bucket=self.bucket_name, Key=part_key
                        )
                        async with get_resp["Body"] as stream:
                            part_data = await stream.read()
                        upload_resp = await self._client.upload_part(
                            Bucket=self.bucket_name,
                            Key=final_key,
                            UploadId=aws_upload_id,
                            PartNumber=idx,
                            Body=part_data,
                        )
                        etag = upload_resp["ETag"]
                    else:
                        raise

                parts_manifest.append({"ETag": etag, "PartNumber": idx})

            resp = await self._client.complete_multipart_upload(
                Bucket=self.bucket_name,
                Key=final_key,
                UploadId=aws_upload_id,
                MultipartUpload={"Parts": parts_manifest},
            )
            etag = resp.get("ETag", "")
            return etag.strip('"')

        except Exception:
            # Abort on any failure
            try:
                await self._client.abort_multipart_upload(
                    Bucket=self.bucket_name,
                    Key=final_key,
                    UploadId=aws_upload_id,
                )
            except Exception:
                logger.warning("Failed to abort AWS multipart upload %s", aws_upload_id)
            raise

    async def delete_parts(self, bucket: str, key: str, upload_id: str) -> None:
        """Delete all temporary part objects for a multipart upload.

        Lists objects under .parts/{upload_id}/ and batch-deletes them
        (up to 1000 per request, matching S3 delete_objects limit).
        """
        prefix = f"{self.prefix}.parts/{upload_id}/"
        paginator = self._client.get_paginator("list_objects_v2")

        async for page in paginator.paginate(Bucket=self.bucket_name, Prefix=prefix):
            contents = page.get("Contents", [])
            if not contents:
                continue

            # Batch delete (max 1000 per call)
            objects = [{"Key": obj["Key"]} for obj in contents]
            await self._client.delete_objects(
                Bucket=self.bucket_name,
                Delete={"Objects": objects, "Quiet": True},
            )

    async def copy_object(
        self,
        src_bucket: str,
        src_key: str,
        dst_bucket: str,
        dst_key: str,
    ) -> str:
        """Copy an object using AWS server-side copy.

        Returns:
            The hex-encoded MD5 ETag of the copied object (quotes stripped).
        """
        src_s3_key = self._s3_key(src_bucket, src_key)
        dst_s3_key = self._s3_key(dst_bucket, dst_key)

        resp = await self._client.copy_object(
            Bucket=self.bucket_name,
            Key=dst_s3_key,
            CopySource={"Bucket": self.bucket_name, "Key": src_s3_key},
        )
        etag = resp.get("CopyObjectResult", {}).get("ETag", "")
        return etag.strip('"')
